### PR TITLE
DBZ-4036 handle SCN gap.

### PR DESCRIPTION
**Pull request by zxpzlp**

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@Naros, could you check out this one?

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnection.java [hunk 1]
```diff
@@ -336,9 +337,11 @@ else if (column.jdbcType() == OracleTypes.NUMBER) {
      * @throws IllegalStateException if the query does not return at least one row
      */
     public Scn getCurrentScn() throws SQLException {
-        return queryAndMap("SELECT CURRENT_SCN FROM V$DATABASE", (rs) -> {
+        return queryAndMap("SELECT CURRENT_SCN,scn_to_timestamp(CURRENT_SCN) FROM V$DATABASE", (rs) -> {
             if (rs.next()) {
-                return Scn.valueOf(rs.getString(1));
+                Scn scn = Scn.valueOf(rs.getString(1));
+                scn.setTime(rs.getObject(2, OffsetDateTime.class));
+                return scn;
```
Comment (Naros) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnection.java:344]: I think I would prefer we break out the `scn_to_timestamp` call into a separate method on the `OracleConnection` class similar to something like as follows:

```
public OffsetDateTime getScnToTimestamp(Scn scn) throws SQLException {
    return queryAndMap("SELECT scn_to_timestamp(" + scn.toString() + ") FROM DUAL", rs -> {
       if (rs.next()) {
         return rs.getObject(1, OffsetDateTime.class);
       }
       throw new DebeziumException("Failed to get SCN as a timestamp");
    });
}
```

So inside `LogMinerStreamingChangeEventSource` method we can use this special method to get the SCN as a time value so we can then do the comparison between the two SCN values easily to determine the time differential.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/Scn.java [hunk 2]
```diff
@@ -27,10 +28,20 @@
 
     private final BigInteger scn;
 
+    private OffsetDateTime time;
```
Comment (Naros) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/Scn.java:31]: Based on the above recommendation, we would remove any changes made to the `Scn` domain class, removing the new time field and the new getting/setter methods.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java [hunk 3]
```diff
@@ -449,20 +451,34 @@ private Scn calculateEndScn(OracleConnection connection, Scn startScn, Scn prevE
...
-                LOGGER.debug("Top SCN calculation resulted in end before start SCN, using current SCN {} as end SCN.", currentScn);
+                LOGGER.debug("Top SCN calculation resulted in end before start SCN, using current SCN {} as end SCN.", currentScn.getFullString());
                 return currentScn;
             }
-            LOGGER.debug("Using Top SCN calculation {} as end SCN.", topScnToMine);
+
+            if (prevEndScn != null) {
+                if (prevEndScn.getTime() != null) {
+                    long seconds = ChronoUnit.SECONDS.between(prevEndScn.getTime(), currentScn.getTime());
+                    if (currentScn.subtract(prevEndScn).longValue() > 10000000 && seconds < 20) {
```
Comment (Naros) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java:471]: Could we extract the SCN gap value, `10000000`, and time variance value, `20`, to configuration options and set these as defaults to those options allowing users to adjust these for their environment?

**Review by Naros — COMMENTED** [review]
Overall this LGTM, I've added some inline recommendations/comments if you could take a look.

**Comment by gunnarmorling** [comment]
@Naros, should this be kept for 1.8? If so, can you add the right label to this PR? Thx!

**Comment by gunnarmorling** [comment]
@Naros, anything left to address here? If not, could you merge it to master?

**Comment by gunnarmorling** [comment]
@Naros, ping?

**Comment by Naros** [comment]
The main outstanding concern here is at startup we have no real way to detect SCN jumps; this PR only addresses the concern at runtime.  I'd like to see if there is a possibility to detect this when the connector is down.

For short outages, we could look at adding the SCN's timestamp to the offsets for `scn` and `commit_scn`.  We could then use these timestamps to compare the window with the current SCN and if within the proximity window, advance forward if we detect an SCN gap jump.

For longer outages, some hours or days, this is where it gets tricky and I'm not sure there is even a possibility for this use case.  I'm inclined to think for this use case, it might be useful to document it and suggest that under the scenario stop the connector and manually adjust the offsets and restart.  @ashulin any thoughts or ideas?

**Comment by Naros** [comment]
> Maybe we can use the timestamp_to_scn function, the logic looks like the following.

@zxpzlp ultimately we have to be careful with the use of `timestamp_to_scn` and `scn_to_timestamp`.  These functions only work with values that are still mapped and retained by the undo retention period.  Once the mapping is discarded, these functions will throw an exception if the mapping of scn to time or time to scn cannot be resolved.

> If we get empty redo log result for several times, it's possible a SCN jump, then we go into the following flow.
We get the timestamp of previous end SCN by using the scn_to_timestamp function first, then add 20 seconds, then use the new timestamp to get the SCN by using the timestamp_to_scn function, if the SCN is much bigger than previous end SCN, we know it's a SCN jump.

Getting the timestamp of the previous end SCN via `scn_to_timestamp` can lead to the problem I mentioned above.  If the connector is down for an extended period of time and the SCN that is calculated as the endpoint based on the batch size is outside the scope of the undo retention period, the function call will fail.

So basically we cannot use the `startScn` nor `endScn` values with these functions at all.  The only value that is guaranteed to be in the undo retention period mapping will be the current scn only.

**Comment by roldanbob** [comment]
> Hi @roldanbob I applied your suggestions in the latest commit [f78422c](https://github.com/debezium/debezium/commit/f78422cd838274a583d2b2d3da6c10c42e7a7605). I also replied to the passage question you had so let me know if this looks good or requires any additional tweaks before I merge.

@Naros Appreciate your response(s!) to that question.  I made one comment in the commit regarding formatting, but otherwise it LGTM.

**Comment by Naros** [comment]
Rebased and applied to master (1.8); will be backported to 1.7 on the first 1.7.1.Final release.
